rust-wrapper/src/msm.rs

```diff
@@ -0,0 +1,76 @@
+use ark_bls12_381::{Fr as ScalarField, G1Affine as GAffine, G1Projective as G};
+use ark_ec::VariableBaseMSM as BaselineVariableBaseMSM;
+use ark_ff::PrimeField;
+use ark_msm::msm::VariableBaseMSM;
+use ark_serialize::CanonicalSerialize;
+use ark_std::log2;
+use haskell_ffi::{from_haskell::marshall_from_haskell_var, to_haskell::marshall_to_haskell_var};
+use utils::{Buffer, RW};
+
+use crate::utils::{self, deserialize_vector_points, deserialize_vector_scalar_field};
+
+#[no_mangle]
+pub extern "C" fn rust_wrapper_multi_scalar_multiplication(
+    points_var: *const u8,
+    points_len: usize,
+    scalars_var: *const u8,
+    scalars_len: usize,
+    out: *mut u8,
+    out_len: &mut usize,
+) {
+    let scalars: Vec<ScalarField> =
+        marshall_from_haskell_var::<RW, Buffer<ScalarField>>(scalars_var, scalars_len, RW).0;
+    let points: Vec<GAffine> =
+        marshall_from_haskell_var::<RW, Buffer<GAffine>>(points_var, points_len, RW).0;
+
+    let r: GAffine = G::msm(&points, &scalars).unwrap().into();
+
+    marshall_to_haskell_var(&r, out, out_len, RW);
+}
+
+const fn get_opt_window_size(k: u32) -> u32 {
```

Comment:
How is this function derived? What is the meaning of these intervals and constants?

Reply:
this is inner function from `arkmsm` crate. 
It is not public, but is used in the `multi_scalar_mul_custom` call. It was probably expected that the wrapper method `multi_scalar_mul` would be called, but `multi_scalar_mul_g1_glv` (one of the options in `multi_scalar_mul`) does not work correctly at the moment

Reply:
Ok! Then a comment with a link to the reference implementation would be a nice addition